e2e/pos_poa_switch_test.go

```diff
@@ -0,0 +1,168 @@
+	ibftOp "github.com/0xPolygon/polygon-sdk/consensus/ibft/proto"
+	"github.com/0xPolygon/polygon-sdk/crypto"
+	"github.com/0xPolygon/polygon-sdk/e2e/framework"
+	"github.com/0xPolygon/polygon-sdk/types"
+	"github.com/stretchr/testify/assert"
+)
+
+// Test scenario:
+// PoA [0, 29]
+// - Check validator set in each validator equals to genesis validator
+// - 3 validators stake to the validator contract after contract deploys at block #10
+// PoS [30, ]
+// - Check validator set in each validator has only 3 validators
+func TestPoAPoSSwitch(t *testing.T) {
+	var (
+		// switch configuration
+		posDeployContractAt = uint64(10)
+		posStartAt          = uint64(30)
+
+		defaultBalance = framework.EthToWei(1000)
+		stakeAmount    = framework.EthToWei(5)
+	)
+
+	ibftManager := framework.NewIBFTServersManager(
+		t,
+		IBFTMinNodes,
+		IBFTDirPrefix,
+		func(i int, config *framework.TestServerConfig) {
+			config.SetSeal(true)
+			config.PremineValidatorBalance(defaultBalance)
+		})
+
+	// Set switch configuration into genesis.json
+	err := ibftManager.GetServer(0).SwitchIBFTType(ibft.PoS, posStartAt, nil, &posDeployContractAt)
+	assert.NoError(t, err)
+
+	// Get server slice
+	servers := make([]*framework.TestServer, 0)
+	for i := 0; i < IBFTMinNodes; i++ {
+		servers = append(servers, ibftManager.GetServer(i))
+	}
+
+	// Get genesis validators
+	genesisValidatorKeys := make([]*ecdsa.PrivateKey, IBFTMinNodes)
+	genesisValidatorAddrs := make([]types.Address, IBFTMinNodes)
+
+	for idx := 0; idx < IBFTMinNodes; idx++ {
+		validatorKey, err := ibftManager.GetServer(idx).Config.PrivateKey()
+		assert.NoError(t, err)
+
+		validatorAddr := crypto.PubKeyToAddress(&validatorKey.PublicKey)
+		genesisValidatorKeys[idx] = validatorKey
+		genesisValidatorAddrs[idx] = validatorAddr
+	}
+
+	// Start servers
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	ibftManager.StartServers(ctx)
+
+	// Test in PoA
+	validateSnapshot := func(
+		ctx context.Context,
+		srv *framework.TestServer,
+		height uint64,
+		expectedValidators []types.Address,
+	) {
+		res, err := srv.IBFTOperator().GetSnapshot(ctx, &ibftOp.SnapshotReq{
+			Number: height,
+		})
+		assert.NoError(t, err)
+
+		snapshotValidators := make([]types.Address, len(res.Validators))
+		for idx, v := range res.Validators {
+			snapshotValidators[idx] = types.StringToAddress(v.Address)
+		}
+
+		assert.ElementsMatch(t, expectedValidators, snapshotValidators)
+	}
+
+	// Check validator set in each validator
+	var wg sync.WaitGroup
+	for i := 0; i < IBFTMinNodes; i++ {
+		wg.Add(1)
+
+		srv := ibftManager.GetServer(i)
+
+		go func(srv *framework.TestServer) {
+			defer wg.Done()
+
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+
+			// every validator should have 4 validators in validator set
+			validateSnapshot(ctx, srv, 1, genesisValidatorAddrs)
+		}(srv)
+	}
+	wg.Wait()
+
+	// Wait until deploying staking contract
```

Comment:
Typo here:

`Wait until the staking contracti s deployed`

Reply:
Fixed at 464cea9821e28327cb3c8e85cda47c21753feec5